### PR TITLE
Allow configuration of the MCP Server description and metadata

### LINKED ISSUE #34262: Allow configuration of the MCP Server description and metadata
During [initialization][mcp-init] the MCP server returns a `serverInfo` object which includes some information about itself, such as a name, version and description. This may be used in UIs to refer to the server, so we should have a sensible generic default and let the user configure it.

Currently, `mcpServer-1.0` [describes itself as a test server][current-desc], which is not appropriate.

The `serverInfo` block is an [`Implementation` object][impl-schema] in the schema.

[mcp-init]: https://modelcontextprotocol.io/specification/2025-11-25/basic/lifecycle#initialization
[current-desc]: https://github.com/OpenLiberty/open-liberty/blob/eb2d574aae91dc568c7327321451501f01ae7e1b/dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java#L421
[impl-schema]: https://modelcontextprotocol.io/specification/2025-11-25/schema#implementation

### PR DESCRIPTION
**Pull request by Yas-Franco**
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

################################################################################################
Resolves https://github.com/OpenLiberty/open-liberty/issues/34262

### CODE REVIEW

**Comment by Yas-Franco** [comment]
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=0c5b1be1-af55-40a0-92dc-5c793ac70900) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

File: dev/io.openliberty.mcp.internal/resources/OSGI-INF/l10n/metatype.properties [hunk 1]
```diff
@@ -9,3 +9,15 @@ moduleName.description=The name of the module in an EAR. Selects the module to w
...
+mcpServerInfoName.description=The name identifier for this MCP server. This display name is used by MCP clients to identify the server.
+
+mcpServerInfoTitle.name=Server title
+mcpServerInfoTitle.description=A human-readable title for this MCP server. Intended for UI and end-user contexts. 
+
+mcpServerInfoVersion.name=Server version
+mcpServerInfoVersion.description=The version string for this MCP server. This value is used to track the server version.
+
+mcpServerInfoDescription.name=Server description
+mcpServerInfoDescription.description=The description string for this MCP server. This provides a description of what this implementation does for clients or servers to provide context about their purpose and capabilities.
```
Comment (Yas-Franco) [inline dev/io.openliberty.mcp.internal/resources/OSGI-INF/l10n/metatype.properties]: Comment from @jantley-ibm in previous pr: 

```
change
This provides a description of what this implementation does for clients or servers to provide context about their purpose and capabilities.
to
This value describes the purpose and capabilities of the implementation for clients or servers.
```

File: dev/io.openliberty.mcp.internal/resources/OSGI-INF/l10n/metatype.properties [hunk 2]
```diff
@@ -9,3 +9,15 @@ moduleName.description=The name of the module in an EAR. Selects the module to w
 
 path.name=Endpoint path
 path.description=The URL path for this MCP endpoint, relative to the module context root. Sets the MCP endpoint path within that module.
+
+mcpServerInfoName.name=Server name
+mcpServerInfoName.description=The name identifier for this MCP server. This display name is used by MCP clients to identify the server.
+
+mcpServerInfoTitle.name=Server title
+mcpServerInfoTitle.description=A human-readable title for this MCP server. Intended for UI and end-user contexts.
```
Comment (Yas-Franco) [inline dev/io.openliberty.mcp.internal/resources/OSGI-INF/l10n/metatype.properties]: Comment from @jantley-ibm in previous pr: 

```
change
this MCP server. Intended
to
this MCP server, intended
```

**Review by habiblawal1 — DISMISSED** [review]
LGTM

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java [hunk 3]
```diff
@@ -16,11 +16,23 @@
...
  */
 public record McpServerConfigProps(boolean stateless,
                                    String moduleName,
                                    String path,
-                                   String servicePid) implements McpConfig {
+                                   String servicePid,
+                                   String mcpServerInfoName,
+                                   String mcpServerInfoTitle,
+                                   String mcpServerInfoVersion,
+                                   String mcpServerInfoDescription) implements McpConfig {
```
Comment (tevans78) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java]: mcpServerInfoDescription is missing from the javadoc

Comment (tevans78) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java]: is it worth wrapping these four into a single MCPServerInfo object?

File: dev/io.openliberty.mcp.internal/resources/OSGI-INF/metatype/metatype.xml [hunk 4]
```diff
@@ -30,6 +30,24 @@
...
+	    id="mcpServerInfoTitle"
+	    required="false"
+	    type="String"
+	    default="Open Liberty MCP Server"/>
+	<AD name="%mcpServerInfoVersion.name"
+	    description="%mcpServerInfoVersion.description"
+	    id="mcpServerInfoVersion"
+	    required="false"
+	    type="String"
+	    default="1.0"/>
```
Comment (tevans78) [inline dev/io.openliberty.mcp.internal/resources/OSGI-INF/metatype/metatype.xml]: Should `<mcpServerInfo>` be a sub-element of `<mcpServer>`? Was there a design for this that I haven't seen?

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java [hunk 5]
```diff
@@ -413,8 +417,13 @@ private void initialize(McpTransport transport) throws IOException {
 
         ServerCapabilities caps = ServerCapabilities.of(new Capabilities.Tools(false));
 
-        // TODO: provide a way for the user to set server info
-        ServerInfo info = new ServerInfo("test-server", "Test Server", "0.1");
+        // Use configured serverInfo with sensible defaults
+        String serverName = mcpConfig.serverInfoName() != null ? mcpConfig.serverInfoName() : io.openliberty.mcp.internal.config.McpServerConfigProps.DEFAULT_SERVER_NAME;
```
Comment (IssacAX123) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java]: you could simplify this using a method from Object class (need to find it)

Comment (IssacAX123) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpServlet.java]: you could try [Objects.requireNonNullElse​(T obj, T defaultObj)](https://docs.oracle.com/javase/9/docs/api/java/util/Objects.html#requireNonNullElse-T-T-) or a helper method

File: dev/io.openliberty.mcp.internal/resources/OSGI-INF/l10n/metatype.properties [hunk 6]
```diff
@@ -9,3 +9,18 @@ moduleName.description=The name of the module in an EAR. Selects the module to w
 
 path.name=Endpoint path
 path.description=The URL path for this MCP endpoint, relative to the module context root. Sets the MCP endpoint path within that module.
+
+mcpServerInfo.name=MCP Server Information
```
Comment (tevans78) [inline dev/io.openliberty.mcp.internal/resources/OSGI-INF/l10n/metatype.properties]: Are these key names correct?

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfig.java [hunk 7]
```diff
@@ -17,4 +19,25 @@ public interface McpConfig {
...
 
     String servicePid();
+
+    String serverInfoName();
+
+    String serverInfoTitle();
+
+    String serverInfoVersion();
+
+    String serverInfoDescription();
```
Comment (tevans78) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfig.java]: Are these methods still needed? Wouldn't you have a `serverInfo()` method instead?

Comment (Yas-Franco) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfig.java]: They have now been removed

File: dev/io.openliberty.mcp.internal/resources/OSGI-INF/l10n/metatype.properties (hunk 6 above)
Comment (Yas-Franco) [reply dev/io.openliberty.mcp.internal/resources/OSGI-INF/l10n/metatype.properties]: Updated

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java [hunk 8]
```diff
@@ -9,18 +9,40 @@
...
                                    String path,
-                                   String servicePid) implements McpConfig {
+                                   String servicePid,
+                                   ServerInfo serverInfo) implements McpConfig {
     public static final String FALLBACK_PATH = "/mcp";
-    public static final McpServerConfigProps DEFAULT_CONFIG = new McpServerConfigProps(false, null, FALLBACK_PATH, null);
+    public static final String DEFAULT_SERVER_NAME = "OpenLiberty MCP Server";
+    public static final String DEFAULT_SERVER_TITLE = "OpenLiberty MCP Server";
+    public static final String DEFAULT_SERVER_VERSION = "1.0.0";
+    public static final String DEFAULT_SERVER_DESCRIPTION = "MCP Server running on OpenLiberty";
```
Comment (tevans78) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java]: Is it better to bake these defaults into the metatype? I'm not sure. Doing so would mean that the default values will show in the generated doc. If you did this, the properties then become required, not optional.

Comment (Yas-Franco) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java]: I think it's better to keep them optional so users can omit the entire <info> element and only provide the fields they want to customize (e.g. just version)

Comment (tevans78) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java]: That's what the default values are for, used if not specified by the user ... but ultimately there is always a value set.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfigurationComponent.java [hunk 9]
```diff
@@ -60,11 +71,50 @@ private void processConfig(Map<String, Object> properties) {
...
+        String infoPid = (String) properties.get("info");
+        if (infoPid == null) {
+            return null;
         }
+
+        Configuration infoConfig;
+        try {
+            infoConfig = configAdmin.getConfiguration(infoPid, null);
+        } catch (IOException e) {
+            // Error accessing persistent config storage
```
Comment (TheoGkoumas) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfigurationComponent.java:101]: Should we trace the error here?

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfigurationComponent.java [hunk 10]
```diff
@@ -60,11 +71,50 @@ private void processConfig(Map<String, Object> properties) {
...
+        }
+
+        Dictionary<String, Object> infoProperties = infoConfig.getProperties();
+        if (infoProperties == null) {
+            // Javadoc indicates this is possible during startup
+            // We don't expect this case, but check it to ensure we don't NPE
+            return null;
+        }
+
+        String name = (String) infoProperties.get("name");
```
Comment (TheoGkoumas) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfigurationComponent.java:118]: I assume that we don't need null handling here, as null keys and null values are not allowed in Dictionary in principle.
Is that correct?

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java (hunk 8 above)
Comment (TheoGkoumas) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java]: fyi another constant here (FALLBACK_PATH = "/mcp") is specified as default on metatype for `path` element.
The actual question here is which of the server elements could be consider as mandatory for the definition.

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/serverinfo/CustomServerInfoTest.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        server.startServer();
+    }
+
+    @AfterClass
+    public static void teardown() throws Exception {
+        server.stopServer();
+    }
+
+    @Test
+    public void testCustomServerInfoInInitializeResponse() throws Exception {
```
Comment (TheoGkoumas) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/serverinfo/CustomServerInfoTest.java]: Maybe consider adding a null handling test method to assert the scenario you don't get an NPE if one of those elements are null (or empty strings)

**Review by TheoGkoumas — CHANGES_REQUESTED** [review]
A couple comments related to null handling and tracing, plus a recommendation to extend test coverage.
Other than that, good definition of the extended server info on the mcp config and accurate unit testing.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java (hunk 8 above)
Comment (Azquelt) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java]: Note that only `name` and `version` are required (and therefore need either a default or validation that the user has configured them). The others are all [optional](https://modelcontextprotocol.org/specification/2025-11-25/schema#implementation).

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java [hunk 12]
```diff
@@ -9,18 +9,40 @@
...
  */
 public record McpServerConfigProps(boolean stateless,
                                    String moduleName,
                                    String path,
-                                   String servicePid) implements McpConfig {
+                                   String servicePid,
+                                   ServerInfo serverInfo) implements McpConfig {
     public static final String FALLBACK_PATH = "/mcp";
-    public static final McpServerConfigProps DEFAULT_CONFIG = new McpServerConfigProps(false, null, FALLBACK_PATH, null);
+    public static final String DEFAULT_SERVER_NAME = "OpenLiberty MCP Server";
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java]: I'm unsure about including `OpenLiberty` in any of the default values because this information is returned to the client. I _think_ we generally avoid identifying the server technology to the end user for security reasons. Please check.

Comment (Azquelt) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java]: (also it would be "Open Liberty" 😛 )

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java (hunk 8 above)
Comment (tevans78) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java]: OK so putting default name and version values into the metatype does seem to be the right thing to do. Then allow the others to be blank.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfigurationComponent.java (hunk 9 above)
Comment (Yas-Franco) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfigurationComponent.java:101]: That would be helpful. I will add that in.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java (hunk 12 above)
Comment (Yas-Franco) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java]: It makes sense to remove references to Open Liberty. I have updated the issue with proposed defaults which don't include it.

**Review by TheoGkoumas — DISMISSED** [review]
Looks good to merge! Nice job.
Maybe you could consider squashing the 8 commits in 1 with a corresponding message.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfigurationComponent.java [hunk 13]
```diff
@@ -60,11 +71,56 @@ private void processConfig(Map<String, Object> properties) {
...
+    private ServerInfo parseServerInfo(Map<String, Object> properties) {
+        String infoPid = (String) properties.get("info");
+        if (infoPid == null) {
+            // No info element configured, return defaults from metatype.xml
+            return McpServerConfigProps.DEFAULT_SERVER_INFO;
         }
+
+        Configuration infoConfig;
+        try {
+            infoConfig = configAdmin.getConfiguration(infoPid, null);
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfigurationComponent.java:99]: Does this get re-run if one of the `info` element's attributes changes?

It feels like you shouldn't have to do this, but I guess the alternative might be having a second component which we probably don't want either.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfig.java [hunk 14]
```diff
@@ -17,4 +19,11 @@ public interface McpConfig {
...
 
     String servicePid();
-}
+
+    /**
+     * Returns the ServerInfo object with the configured values (or defaults)
+     *
+     * @return ServerInfo with name, title, version, and description
+     */
+    ServerInfo getServerInfo();
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfig.java]: rename to `serverInfo`?

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/serverinfo/CustomServerInfoTest.java [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+    @BeforeClass
+    public static void setup() throws Exception {
+        WebArchive war = ShrinkWrap.create(WebArchive.class, "customServerInfoTest.war")
+                                   .addPackage(BasicTools.class.getPackage());
+
+        ShrinkHelper.exportAppToServer(server, war, SERVER_ONLY);
+
+        server.startServer();
+    }
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/serverinfo/CustomServerInfoTest.java:72]: We need to test having multiple applications deployed, each with separate config.

Could you combine this test with `PartialServerInfoTest` and deploy both apps on the same server at the same time and check that each have the correct initialization response ?

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfigurationComponent.java (hunk 13 above)
Comment (Yas-Franco) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpConfigurationComponent.java:99]: Issue raised https://github.com/OpenLiberty/open-liberty/issues/34936

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/config/McpServerConfigProps.java:47]: This is ok, though as you point out it should not happen.

Another alternative would be to validate that that it's not null in the constructor.

We're not terribly good at saying which things can be null and which can't. In general we should document when things can be `null` (and what that means) and assume it's not `null` the rest of the time.

**Review by Azquelt — APPROVED** [review]
